lib/src/network/https_openssl.rs

```diff
@@ -239,11 +239,11 @@ impl TlsClient {
     }
   }
 
-  fn front_socket(&self) -> &TcpStream {
+  fn front_socket(&self) -> Option<&TcpStream> {
     match unwrap_msg!(self.protocol.as_ref()) {
-      &State::Handshake(ref handshake) => panic!("we should not need to access the front socket now"),
-      &State::Http(ref http)           => http.front_socket(),
-      &State::WebSocket(ref pipe)      => pipe.front_socket(),
+      &State::Handshake(ref handshake) => handshake.front.as_ref(),
```

Comment:
this might not work because when we call `readable()` in the initial state, that field is emptied and the socket passed to `ssl.accept`:

https://github.com/sozu-proxy/sozu/blob/master/lib/src/network/protocol/openssl.rs#L48-L50

To get access to the socket from the handshake protocol, there should be a method of the handshake protocol that would look at the current state and:

- if `TlsState::Initial`, do `self.front.as_ref()`
- if `TlsState::Handshake`, do `self.mid.as_ref().map(|m| m.get_ref())` (cf https://docs.rs/openssl/0.10.5/openssl/ssl/struct.MidHandshakeSslStream.html#method.get_ref )
- if `TlsState::Established`, do `self.stream.as_ref().map(|m| m.get_ref())` ( cf https://docs.rs/openssl/0.10.5/openssl/ssl/struct.SslStream.html#method.get_ref ) although whenever we reach that state, we perform an upgrade immediately, so it might not be necessary
- if `TlsState::Error` we should have stored the resulting error somewhere since it can contain a `MidHandshakeSslStream` with access to the socket: https://docs.rs/openssl/0.10.5/openssl/ssl/enum.HandshakeError.html